### Data Cleaning Project - John Cavers
The aim of this project is to clean the data and analyze the included used car listings. I will also become familiar with some of the unique benefits jupyter notebook provides for pandas.

##### Data Background
The dataset on Ebay car listings was originally scraped and uploaded to Kaggle. We've made a few modifications from the original dataset that was uploaded to Kaggle.  We sampled 50,000 data points from the full dataset, to ensure your code runs quickly in our hosted environment. We dirtied the dataset a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)

> The data dictionary provided with data is as follows:

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.

Dataquest: Guided Project: Exploring Ebay Car Sales Data

In [169]:
# import Python Pandas library
import pandas as pd
# import Python Numpy library
import numpy as np


In [170]:
# read csv file into Panda dataframe
autos = pd.read_csv('autos.csv', encoding = "Latin-1")

In [171]:
autos 
# A neat feature of jupyter notebook is its ability to render the first
# few and last few values of any pandas object

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [172]:
# run info command to get information on the dataframe
autos.info()
# run head command to look at first five rows
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

***Initial Data Analysis:*** Looking at the above information I see there are 5 columns with null values, though none in excess of 20% null values. Five columns are integers and 15 are strings.   The column names use camelcase instead of Pythons preferred snakecase so I can't just replace spaces with underscores. There is definitely some data to cleanup and perhaps some rows/columns to drop from the data set.  I need to explore further to determine.

In [173]:
# in this step I will update the column names
# print current column names so we can see before cleanup
print(autos.columns)
# create dictionary with column name mapping to change to
column_names_mapping = {"dateCrawled" : "date_crawled", 
  "name" : "name", 
  "seller" : "seller", 
  "offerType" : "offer_type", 
  "price" : "price", 
  "abtest": "ab_test",
  "vehicleType" :"vehicle_type", 
  "yearOfRegistration" : "registration_year", 
  "gearbox" : "gear_box", 
  "powerPS" : "power_ps", 
  "model" : "model",
  "odometer" : "odometer", 
  "monthOfRegistration" : "registration_month", 
  "fuelType" : "fuel_type", 
  "brand" : "brand",
  "notRepairedDamage" : "unrepaired_damage", 
  "dateCreated" : "ad_created", 
  "nrOfPictures" : "nr_of_pictures",
  "postalCode" : "postal_code",
  "lastSeen": "last_seen"}

# run dataframe.rename command to update column names passing dictionary as an argument
autos.rename(column_names_mapping, axis = 1, inplace = True)
print('\n' * 2) # print two blank lines
# print column names after the above line of code was excetued
print(autos.columns)
# look at impact on data set
autos.head()

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')



Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

***Above Code Goal*** - the intention was to make the column names more readable to users. Further I wanted to adopt standard Python naming conventions to make it easier for other developers and analysts to follow this work.

****Next Step: Data Exploration*** Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for: - Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis. - Examples of numeric data stored as text which can be cleaned and converted.

The following methods are helpful for exploring the data: - DataFrame.describe() (with include='all' to get both categorical and numeric columns) - Series.value_counts() and Series.head() if any columns need a closer look.

In [174]:
# use dataframe.describe to look at descriptive statistics for all columns
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year gear_box      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [175]:
# use value.counts and head command to explore each series in the dataframe
# take notes and write in markdown cell findings and suggestions
autos["fuel_type"].value_counts()
# autos["odometer"].head(55)

benzin     30107
diesel     14567
lpg          691
cng           75
hybrid        37
andere        22
elektro       19
Name: fuel_type, dtype: int64

***Data Exploration Findings***
- Seller: Change word to 'private' from 'privat'. Only 2 unique values. One has only one record.  ___Candidate for being dropped___.

- Name: We could stop it after 1-2 underscores and remove the rest. There are a lot of single entries with extra extraneous comments. First field is the maker, second field is the model.

- Offer_type: Only two entries. One has only one record.  ___Candidate for being dropped___

- Price: Data looks good, but there are 1421 records at 0 value. A comma and dollar sign will need to be stripped out to make numeric. ___Change from string to integer___

- ab_test:  looks fine. 2 values evenly split

- vehicle_type: 7 unique values. Some have foreign type language like kleinwagen.  Appears German.

- Registration_year: Bulk of data looks fine. There are small samples of years not possible like 1111, 6200, etc

- Gear_box: This is actually transmission type.  Only two values.  We can update them to more English names as well as change the column name

-  power_ps: I am not sure what that means.  Values are all numbers in the data.  5500 records have a 0 record. ___Needs more investigation___

- Model: looks similar to the second word in the name column

- Odometer: has a “KM” on the end.  We can remove that and make this an integer. ___Change from string to integer___

- Registration_month: values is 0-12.  There should not be any 0, but there are 4075 records.

- Fuel_type: looks fine though it is a foreign way of saying it.  Only 7 unique values

- Brand: looks fine through there are some brands I have not heard before, but they look to be reasonable.

- unrepaired_damage: - only two values and it is in German.  You have No and And as values. I believe this should be Yes and No.

- ad_created: is a date formatted as 2016-04-03 00:00:00.  We can clean that up as all have no time and only a calendar date.

- nr_of_pictures: have no values as all are zero.  __Candidate for being dropped___

- postal_code: for the most part looks good.  I saw one that was only four digits in length

- last_seen: looks like a full calendar and timestamp.  This one is good.


In [176]:
# in autos column, strip out sring characters and convert to integer
# look at data and determine characters to strip out
print(autos["price"].head())
# following has to be replaced "$", "," in the data
autos["price"] = autos["price"].str.replace("$", "").str.replace(",", "").astype(int)
# check work below and confirm int64 as dtype
print(autos["price"].head())

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object
0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64


In [177]:
# in odometer column, strip out sring characters and convert to integer
# look at unique values
autos["odometer"].value_counts()
# comma and km need to be stripped out
print(autos["odometer"].head())
autos["odometer"] = autos["odometer"].str.replace(",", "").str.replace("km", "").astype(int)
# confirm change is to dtype int64
print(autos["odometer"].head())
# change column name to odometer_km
autos.rename({"odometer" : "odometer_km"}, axis = 1, inplace = True)
# confirm column name change and string to integer changes
autos.info()

0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object
0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 50000 non-null int64
ab_test               50000 non-null object
vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gear_box              47320 non-null object
power_ps              50000 non-null int64
model                 47242 non-null object
odometer_km           50000 non-null int64
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null objec

In [178]:
# investigating odometer_km and price columns in more detail
autos["odometer_km"].unique().shape
# 13 unique values exist for odometer_km
# look at min/max/median for the series
autos["odometer_km"].describe()
autos["odometer_km"].value_counts().sort_index(ascending=True)
# I do not see any real issues with this column's data
# I will now look at the price column
autos["price"].unique().shape
# there are 2357 unique prices values
autos["price"].describe()
autos["price"].value_counts().sort_index(ascending=True).head(10)
# there looks to be some unrealistic values in price column.  I should drop all rows where the price is greater than 150,000 or less than 500.

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price, dtype: int64

In [179]:
# remove outliers in price less than 500 or greater than 150000
# create a boolean argument to remove those rows from dataframe
autos = autos[autos["price"].between(500,150000)]
autos["price"].value_counts().sort_index(ascending=True).tail()


130000    1
135000    1
137999    1
139997    1
145000    1
Name: price, dtype: int64

___Data Justification___ In my analysis I chose to remove rows of data where the price appeared out of the norm for what I was expecting. There were a number of zero records and some close to one million. I decided to modify the data set and only keep those rows which were greater than 500 and less than 150000.  

In [180]:
# look at all the date columns to see if they need formatting to enable statiscal analysis
autos[['date_crawled','ad_created','last_seen', 'registration_month', 'registration_year']][0:5]

date_crawled           ad_created            last_seen  \
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54   
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08   
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37   
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28   
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50   

   registration_month  registration_year  
0                   3               2004  
1                   6               1997  
2                   7               2009  
3                   6               2007  
4                   7               2003

In [181]:
# date_crawled, ad_created, and last_seen are all strings and will need to be converted.
# first I will research date_crawled
# print(autos["date_crawled"].str[:10])
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025534
2016-03-06    0.014131
2016-03-07    0.036205
2016-03-08    0.033165
2016-03-09    0.032921
2016-03-10    0.032722
2016-03-11    0.033032
2016-03-12    0.037336
2016-03-13    0.015529
2016-03-14    0.036316
2016-03-15    0.034031
2016-03-16    0.029350
2016-03-17    0.031147
2016-03-18    0.012867
2016-03-19    0.034763
2016-03-20    0.038068
2016-03-21    0.037735
2016-03-22    0.032988
2016-03-23    0.032389
2016-03-24    0.028973
2016-03-25    0.031102
2016-03-26    0.032655
2016-03-27    0.031147
2016-03-28    0.034807
2016-03-29    0.033276
2016-03-30    0.033299
2016-03-31    0.031679
2016-04-01    0.033920
2016-04-02    0.035783
2016-04-03    0.038822
2016-04-04    0.036626
2016-04-05    0.013155
2016-04-06    0.003172
2016-04-07    0.001353
Name: date_crawled, dtype: float64

___Date Crawled Observations___ The data covers from 03/05/2016 to 04/07/2016 so roughly 32 days.  Looking at the percent distribution, most values look to be around three percent with a few days around one percent.  I don't see an overweighting on one day which could skew results.  These dates look good.  

In [182]:
# date_crawled, ad_created, and last_seen are all strings and will need to be converted.
# second I will research ad_created
# print(autos["date_crawled"].str[:10])
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False)

2016-04-07    0.001198
2016-04-06    0.003261
2016-04-05    0.011891
2016-04-04    0.037003
2016-04-03    0.039044
2016-04-02    0.035473
2016-04-01    0.033853
2016-03-31    0.031746
2016-03-30    0.033165
2016-03-29    0.033254
2016-03-28    0.034851
2016-03-27    0.031102
2016-03-26    0.032655
2016-03-25    0.031258
2016-03-24    0.028950
2016-03-23    0.032212
2016-03-22    0.032744
2016-03-21    0.037957
2016-03-20    0.038201
2016-03-19    0.033631
2016-03-18    0.013488
2016-03-17    0.030814
2016-03-16    0.029838
2016-03-15    0.033809
2016-03-14    0.034896
2016-03-13    0.016971
2016-03-12    0.037114
2016-03-11    0.033343
2016-03-10    0.032456
2016-03-09    0.033010
                ...   
2016-02-21    0.000044
2016-02-20    0.000044
2016-02-19    0.000067
2016-02-18    0.000044
2016-02-17    0.000022
2016-02-16    0.000022
2016-02-14    0.000044
2016-02-12    0.000044
2016-02-11    0.000022
2016-02-09    0.000022
2016-02-08    0.000022
2016-02-07    0.000022
2016-02-05 

___Ad Created Observations___ The data covers from 06/11/2015 to 04/07/2016 so about ten months. Starting on 03/09/2016 the dates run about three percent per day. Prior to that day, the distribution numbers are very low.  I will look further to see if I should drop those rows from the data set.

In [183]:
# date_crawled, ad_created, and last_seen are all strings and will need to be converted.
# Last I will research last_seen
# print(autos["last_seen"].str[:10])
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False)

2016-04-07    0.134148
2016-04-06    0.225370
2016-04-05    0.126561
2016-04-04    0.024314
2016-04-03    0.024935
2016-04-02    0.024869
2016-04-01    0.022872
2016-03-31    0.023449
2016-03-30    0.024159
2016-03-29    0.021363
2016-03-28    0.020520
2016-03-27    0.015440
2016-03-26    0.016461
2016-03-25    0.018590
2016-03-24    0.019544
2016-03-23    0.018413
2016-03-22    0.021253
2016-03-21    0.020654
2016-03-20    0.020410
2016-03-19    0.015418
2016-03-18    0.007387
2016-03-17    0.027686
2016-03-16    0.016150
2016-03-15    0.015684
2016-03-14    0.012290
2016-03-13    0.008874
2016-03-12    0.023915
2016-03-11    0.012046
2016-03-10    0.010293
2016-03-09    0.009450
2016-03-08    0.007010
2016-03-07    0.005213
2016-03-06    0.004171
2016-03-05    0.001087
Name: last_seen, dtype: float64

___Last Seen Observations___ The data covers from 03/05/2016 to 04/07/2016 so roughly 32 days.  This follows the same pattern as date_crawled.  Looking at the percent distribution, most values look to be in a range of 1.5-2.5 percent with a few outliers. I don't see an overweighting on one day which could skew results.  These dates look good.  

In [184]:
# I will next look at the registration_year in more detail
print(autos["registration_year"].describe())
# when I ran describe I saw a max value of 9999 which is not a valid year
autos["registration_year"].value_counts().sort_index()

count    45077.000000
mean      2005.062182
std         89.671642
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


1000       1
1001       1
1910       2
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       1
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      19
1961       6
1962       4
1963       8
1964      11
1965      17
1966      21
        ... 
2000    2714
2001    2519
2002    2419
2003    2657
2004    2683
2005    2864
2006    2666
2007    2269
2008    2205
2009    2078
2010    1586
2011    1617
2012    1304
2013     795
2014     651
2015     360
2016     946
2017    1296
2018     462
2019       1
2800       1
4100       1
4500       1
4800       1
5000       2
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, Length: 93, dtype: int64

___Registration Year Observations___ I see values from 2800-9999 which are not accurate values. Also I see 1000-1001 which both are not accurate.  There are very low row counts in these inaccuracies. One thought I can exclude them from the data set.  Another possiblity is that I could update them to the mean value which is 2005. I checked that cars were created in 1885 so it is feasible those early 1900's records are correct. Also, this data set is from 2016 so the records after 2016 look questionable. My conclusion is to remove those rows from the data set.   

___Next Step Action - Only Keep years 1910-2016 data___

In [185]:
# below line of code removes rows in data set which fall outside the 1900-2016 year range
autos = autos[autos["registration_year"].between(1900,2016)]
# running below command to make sure rows were removed correctly
# autos["registration_year"].value_counts().sort_index()
autos["registration_year"].value_counts(normalize=True, dropna=False).sort_index(ascending=False)

2016    0.021846
2015    0.008314
2014    0.015034
2013    0.018359
2012    0.030113
2011    0.037342
2010    0.036626
2009    0.047987
2008    0.050920
2007    0.052398
2006    0.061566
2005    0.066139
2004    0.061959
2003    0.061358
2002    0.055862
2001    0.058171
2000    0.062675
1999    0.059395
1998    0.046764
1997    0.034917
1996    0.024756
1995    0.019860
1994    0.011547
1993    0.007852
1992    0.007043
1991    0.006766
1990    0.006328
1989    0.003579
1988    0.002979
1987    0.001570
          ...   
1968    0.000600
1967    0.000600
1966    0.000485
1965    0.000393
1964    0.000254
1963    0.000185
1962    0.000092
1961    0.000139
1960    0.000439
1959    0.000139
1958    0.000092
1957    0.000046
1956    0.000092
1955    0.000046
1954    0.000046
1953    0.000023
1952    0.000023
1951    0.000046
1950    0.000023
1948    0.000023
1943    0.000023
1941    0.000046
1939    0.000023
1938    0.000023
1937    0.000092
1934    0.000046
1931    0.000023
1929    0.0000

___Updated Registration Year Observations___ Removing the bad date values dramatically change percentile distribution.  Now there are several years with 5-6 percent which were just over 3 percent prior to the cleanup. This looks much better for calculating statistics.

In [186]:
# let me explore the brand column
autos["brand"].value_counts(normalize=True, dropna=False)

volkswagen        0.211995
bmw               0.115004
mercedes_benz     0.102256
opel              0.100386
audi              0.090525
ford              0.065192
renault           0.043923
peugeot           0.029374
fiat              0.023532
seat              0.017897
skoda             0.017158
smart             0.015126
nissan            0.015011
mazda             0.015011
citroen           0.014156
toyota            0.013417
hyundai           0.010161
sonstige_autos    0.009653
mini              0.009376
volvo             0.009376
honda             0.007805
mitsubishi        0.007736
kia               0.007344
alfa_romeo        0.006628
porsche           0.006143
chevrolet         0.006027
suzuki            0.005912
chrysler          0.003603
dacia             0.002840
jeep              0.002448
land_rover        0.002263
daihatsu          0.002171
subaru            0.001986
saab              0.001663
jaguar            0.001617
daewoo            0.001339
rover             0.001247
t

___Top Car Brands___ Looking at the above results, I plan to focus on teh top ten brands based on greatest percentile distribution.  Those brands are:  Volkswagen, BMW, Mercedes Benz, Ford, Opel, Audi, Fiat, Seat, Renault, and Puegot.  I now want to explore those brands in more detail by finding out the mean price for each one of them.  

In [187]:
# create a variable using value counts to select the top ten brands
selected_brand = autos['brand'].value_counts(normalize=True).index[:10]
# pass the above variable into autos to select only those records from the data frame
mean_price_autos=autos.loc[autos['brand'].isin(selected_brand)]
# create variable that calculates the mean of the data frame
selected=mean_price_autos.groupby(['brand']).mean()
# execute selected to see the mean results
selected

price  registration_year    power_ps    odometer_km  \
brand                                                                      
audi           9571.457398        2004.361735  163.023214  128941.326531   
bmw            8447.069880        2003.120482  169.620482  132928.714859   
fiat           3256.152110        2003.689892   73.913641  114416.094210   
ford           4291.666312        2003.247609  107.037549  123494.509387   
mercedes_benz  8691.820235        2002.091915  153.401084  131058.039747   
opel           3394.039568        2002.731079  101.912583  128012.422360   
peugeot        3360.920597        2004.133648   96.096698  126073.113208   
renault        2819.059411        2003.383807   86.167718  126351.209253   
seat           4810.883871        2005.009032  102.118710  120058.064516   
volkswagen     5783.622985        2002.835076  102.481155  128234.749455   

               registration_month  nr_of_pictures   postal_code  
brand                                                            
audi                     6.177041             0.0  54981.445663  
bmw                      6.094779             0.0  55233.355422  
fiat                     5.838077             0.0  55158.302257  
ford                     5.878144             0.0  50325.467942  
mercedes_benz            6.151536             0.0  51569.003613  
opel                     5.753853             0.0  50304.378422  
peugeot                  5.844340             0.0  53377.887579  
renault                  5.870137             0.0  48483.603049  
seat                     6.012903             0.0  52433.621935  
volkswagen               5.881264             0.0  49687.007952

___Mean Price Findings___ There is definitely a difference in price.  BMW, Mercedes Benz, and Audi are far away the more expensive cars in the data set.  Fiat, Opel, Puegot, and Renault are on the lower side of pricing.  Ford, Volkswagen, and Seat price is in the middle.

In [188]:
# create empty dictionaries for mean price and mean mileage
mean_price={}
mean_mileage={}
# populate the dictionaries with values from the selected dataframe
mean_price=selected.iloc[:,0]
mean_mileage=selected.iloc[:,3]
#print(mean_price)


In [189]:
# map mean price and mean mileage series into the autos dataframe
autos['mean_price']=autos['brand'].map(mean_price)
autos['mean_mileage']=autos['brand'].map(mean_mileage)
# review to see if it is as expected
autos.head(25)

date_crawled                                               name  \
0   2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1   2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2   2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3   2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5   2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
7   2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
9   2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
12  2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13  2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14  2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15  2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
17  2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
19  2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20  2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21  2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22  2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23  2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24  2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
26  2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
28  2016-03-19 21:56:19                                      MINI_Cooper_D   
29  2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
31  2016-03-14 16:53:09                          Noch_gut_erhaltenen_C_320   
33  2016-03-15 12:25:26  Audi_A3_2.0_TDI_Sportback_DPF_Kupplung_defekt_...   
34  2016-04-05 07:55:07                            Bmw_530d__navi_tuv03/18   

    seller offer_type  price  ab_test vehicle_type  registration_year  \
0   privat    Angebot   5000  control          bus               2004   
1   privat    Angebot   8500  control    limousine               1997   
2   privat    Angebot   8990     test    limousine               2009   
3   privat    Angebot   4350  control   kleinwagen               2007   
4   privat    Angebot   1350     test        kombi               2003   
5   privat    Angebot   7900     test          bus               2006   
7   privat    Angebot   1990  control    limousine               1998   
9   privat    Angebot    590  control          bus               1997   
12  privat    Angebot   5299  control   kleinwagen               2010   
13  privat    Angebot   1350  control    limousine               1999   
14  privat    Angebot   3999     test   kleinwagen               2007   
15  privat    Angebot  18900     test        coupe               1982   
17  privat    Angebot   5500     test        coupe               1990   
19  privat    Angebot   4150  control          suv               2004   
20  privat    Angebot   3500     test        kombi               2003   
21  privat    Angebot  41500     test       cabrio               2004   
22  privat    Angebot  25450  control       cabrio               2015   
23  privat    Angebot   7999  control          bus               2010   
24  privat    Angebot  48500  control    limousine               2014   
26  privat    Angebot    777  control   kleinwagen               1992   
28  privat    Angebot   5250  control   kleinwagen               2007   
29  privat    Angebot   4999     test        kombi               2004   
31  privat    Angebot   2850     test        kombi               2002   
33  privat    Angebot   4800     test        kombi               2006   
34  privat    Angebot   2600     test        kombi               2000   

     gear_box  power_ps      ...       registration_month  fuel_type  \
0     manue

___Data Findings after adding Mean Price & Mean Mileage___ the original theory holds regarding brand mean prices after adding mean mileage.  The mileage numbers are not that much different.  Had they been different that could have changed mean prices.  Thus, the data shows BMW, Audi, and Mercedez Benz are clearly the most expensive of the car brands.

In [190]:
# dropping the number of pictures column as that does not add anything to the data set
autos = autos.drop(columns = "nr_of_pictures")
autos.head(3)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offer_type  price  ab_test vehicle_type  registration_year   gear_box  \
0    Angebot   5000  control          bus               2004    manuell   
1    Angebot   8500  control    limousine               1997  automatik   
2    Angebot   8990     test    limousine               2009    manuell   

   power_ps      ...       odometer_km  registration_month  fuel_type  \
0       158      ...            150000                   3        lpg   
1       286      ...            150000                   6     benzin   
2       102      ...             70000                   7     benzin   

        brand unrepaired_damage           ad_created postal_code  \
0     peugeot              nein  2016-03-26 00:00:00       79588   
1         bmw              nein  2016-04-04 00:00:00       71034   
2  volkswagen              nein  2016-03-26 00:00:00       35394   

             last_seen   mean_price   mean_mileage  
0  2016-04-06 06:45:54  3360.920597  126073.113208  
1  2016-04-06 14:45:08  8447.069880  132928.714859  
2  2016-04-06 20:15:37  5783.622985  128234.749455  

[3 rows x 21 columns]

There is another notebook example here which uses some different techniques [Advanced Notebook Example](https://github.com/Zubair063/Data-Science-project/blob/master/exploring%20ebay%20car%20sell/Basics.ipynb)